# Generated by Django 3.2.6 on 2021-08-17 02:49

from django.db import migrations, models


class Migration(migrations.Migration):

    initial = True

    dependencies = [
    ]

    operations = [
        migrations.CreateModel(
            name='Input',
            fields=[
                ('id', models.AutoField(auto_created=True, primary_key=True, serialize=False, verbose_name='ID')),
                ('name', models.CharField(default='Test_name', max_length=100, verbose_name='Your submission')),
                ('upfile_fastq', models.FileField(blank=True, null=True, upload_to='users_file/', verbose_name='Upload your fastq files (all-in zip format, group by study group required)')),
                ('gene_count_method', models.CharField(choices=[('Rsubread', 'Rsubread/featureCounts (Liao et. al. 2014) for gene counts'), ('HTSeq', 'HTSeq/htseq-counts (Ander et. al. 2014) for gene counts'), ('Salmon', 'Salmon (Patro et. al. 2017) for transcriptome counts')], default='Rsubread', max_length=100)),
                ('Differential_expression_method', models.CharField(choices=[('DESeq2', 'DESeq2 (Love et. al. 2014)'), ('limma', 'limma (Ritchie et. al. 2015)'), ('edgeR', 'edgeR (Robinson et. al. 2010)')], default='DESeq2', max_length=100)),
                ('reference_genes', models.CharField(choices=[('human', 'Human hg38 (Homo sapiens)'), ('mouse', 'Mouse (Mus musculus)'), ('rat', 'Rat (Rattus norvegicus)'), ('zebrafish', 'Zebrafish (Danio rerio)'), ('celegans', 'Caenorhabditis elegans')], default='human', max_length=100)),
                ('reference_group', models.CharField(default='first group', help_text='Your reference group (must be one of your subfolder name)', max_length=100, verbose_name="Reference group (reference's sub-directory name)")),
                ('study_group', models.CharField(default='second group', max_length=100, verbose_name="Study group (study groups's sub-directory name)")),
                ('NumberOfTopGene', models.IntegerField(default=30, verbose_name='Number of top variance genes (For Gene Ontology)')),
                ('readCountMinThreshold', models.IntegerField(default=10, verbose_name='readCountMinThreshold (Optional)')),
                ('lfcThreshold', models.FloatField(default=1, verbose_name='lfcThreshold (Optional)')),
                ('adjPValueThreshold', models.FloatField(default=0.05, verbose_name='adjPValueThreshold (Optional)')),
                ('cluster_by_replica', models.CharField(choices=[('Yes', 'Yes'), ('No', 'No')], default='Yes', max_length=6, verbose_name='Cluster replicate')),
                ('pathway_ID', models.CharField(default='hsa04110', max_length=10, verbose_name='KEGG id (for pathview)')),
            ],
        ),
    ]
